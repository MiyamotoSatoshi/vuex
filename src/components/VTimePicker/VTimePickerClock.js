import '../../stylus/components/_time-picker-clock.styl'

// Mixins
import Colorable from '../../mixins/colorable'
import Themeable from '../../mixins/themeable'

/* @vue/component */
export default {
  name: 'v-time-picker-clock',

  mixins: [
    Colorable,
    Themeable
  ],

  props: {
    allowedValues: Function,
    double: Boolean,
    format: {
      type: Function,
      default: val => val
    },
    max: {
      type: Number,
      required: true
    },
    min: {
      type: Number,
      required: true
    },
    scrollable: Boolean,
    rotate: {
      type: Number,
      default: 0
    },
    size: {
      type: [Number, String],
      default: 270
    },
    step: {
      type: Number,
      default: 1
    },
    value: Number
  },

  data () {
    return {
      defaultColor: 'accent',
      inputValue: this.value,
      isDragging: false,
      valueOnMouseDown: null,
      valueOnMouseUp: null
    }
  },

  computed: {
    count () {
      return this.max - this.min + 1
    },
    innerRadius () {
      return this.radius - Math.max(this.radius * 0.4, 48)
    },
    outerRadius () {
      return this.radius - 4
    },
    roundCount () {
      return this.double ? (this.count / 2) : this.count
    },
    degreesPerUnit () {
      return 360 / this.roundCount
    },
    degrees () {
      return this.degreesPerUnit * Math.PI / 180
    },
    radius () {
      return this.size / 2
    },
    displayedValue () {
      return this.value == null ? this.min : this.value
    }
  },

  watch: {
    value (value) {
      this.inputValue = value
    }
  },

  methods: {
    wheel (e) {
      e.preventDefault()

      const delta = Math.sign(e.wheelDelta || 1)
      let value = this.displayedValue
      do {
        value = value + delta
        value = (value - this.min + this.count) % this.count + this.min
      } while (!this.isAllowed(value) && value !== this.displayedValue)

      if (value !== this.displayedValue) {
        this.update(value)
      }
    },
    handScale (value) {
      return this.double && (value - this.min >= this.roundCount) ? (this.innerRadius / this.radius) : (this.outerRadius / this.radius)
    },
    isAllowed (value) {
      return !this.allowedValues || this.allowedValues(value)
    },
    genValues () {
      const children = []

      for (let value = this.min; value <= this.max; value = value + this.step) {
        const classes = {
          active: value === this.displayedValue,
          disabled: !this.isAllowed(value)
        }

        children.push(this.$createElement('span', {
          'class': this.addBackgroundColorClassChecks(classes, value === this.value ? this.computedColor : null),
          style: this.getTransform(value),
          domProps: { innerHTML: `<span>${this.format(value)}</span>` }
        }))
      }

      return children
    },
    genHand () {
      const scale = `scaleY(${this.handScale(this.displayedValue)})`
      const angle = this.rotate + this.degreesPerUnit * (this.displayedValue - this.min)

      return this.$createElement('div', {
        staticClass: 'v-time-picker-clock__hand',
        'class': this.value == null ? {} : this.addBackgroundColorClassChecks(),
        style: {
          transform: `rotate(${angle}deg) ${scale}`
        }
      })
    },
    getTransform (i) {
      const { x, y } = this.getPosition(i)
      return { transform: `translate(${x}px, ${y}px)` }
    },
    getPosition (value) {
      const radius = (this.radius - 24) * this.handScale(value)
      const rotateRadians = this.rotate * Math.PI / 180
      return {
        x: Math.round(Math.sin((value - this.min) * this.degrees + rotateRadians) * radius),
        y: Math.round(-Math.cos((value - this.min) * this.degrees + rotateRadians) * radius)
      }
    },
    onMouseDown (e) {
      e.preventDefault()

      this.valueOnMouseDown = null
      this.valueOnMouseUp = null
      this.isDragging = true
      this.onDragMove(e)
    },
    onMouseUp () {
      this.isDragging = false
      if (this.valueOnMouseUp !== null && this.isAllowed(this.valueOnMouseUp)) {
        this.$emit('change', this.valueOnMouseUp)
      }
    },
    onDragMove (e) {
      e.preventDefault()
      if (!this.isDragging && e.type !== 'click') return

      const { width, top, left } = this.$refs.clock.getBoundingClientRect()
      const { clientX, clientY } = 'touches' in e ? e.touches[0] : e
      const center = { x: width / 2, y: -width / 2 }
      const coords = { x: clientX - left, y: top - clientY }
      const handAngle = Math.round(this.angle(center, coords) - this.rotate + 360) % 360
      const insideClick = this.double && this.euclidean(center, coords) < (this.outerRadius + this.innerRadius) / 2 - 16
      const value = Math.round(handAngle / this.degreesPerUnit) +
        this.min + (insideClick ? this.roundCount : 0)

      // Necessary to fix edge case when selecting left part of max value
      let newValue
      if (handAngle >= (360 - this.degreesPerUnit / 2)) {
        newValue = insideClick ? this.max : this.min
      } else {
        newValue = value
      }

      if (this.isAllowed(value)) {
        if (this.valueOnMouseDown === null) {
          this.valueOnMouseDown = newValue
        }
        this.valueOnMouseUp = newValue
        this.update(newValue)
      }
    },
    update (value) {
      if (this.inputValue !== value) {
        this.inputValue = value
        this.$emit('input', value)
      }
    },
    euclidean (p0, p1) {
      const dx = p1.x - p0.x
      const dy = p1.y - p0.y

      return Math.sqrt(dx * dx + dy * dy)
    },
    angle (center, p1) {
      const value = 2 * Math.atan2(p1.y - center.y - this.euclidean(center, p1), p1.x - center.x)
      return Math.abs(value * 180 / Math.PI)
    }
  },

  render () {
    const data = {
      staticClass: 'v-time-picker-clock',
      class: {
        'v-time-picker-clock--indeterminate': this.value == null,
        ...this.themeClasses
      },
      on: {
        mousedown: this.onMouseDown,
        mouseup: this.onMouseUp,
        mouseleave: () => (this.isDragging && this.onMouseUp()),
        touchstart: this.onMouseDown,
        touchend: this.onMouseUp,
        mousemove: this.onDragMove,
        touchmove: this.onDragMove
      },
      style: {
        height: `${this.size}px`,
        width: `${this.size}px`
      },
      ref: 'clock'
    }

    this.scrollable && (data.on.wheel = this.wheel)

    return this.$createElement('div', data, [
      this.genHand(),
      this.genValues()
    ])
  }
}
