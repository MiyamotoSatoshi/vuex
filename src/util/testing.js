import Vue from 'vue'
import load from '~util/load'
import { mount, shallow } from 'avoriaz'

export function test(name, cb) {
  Vue.prototype.$vuetify = { load }

  beforeEach(() => {
    rafPolyfill(window)
  })

  describe(name, () => cb({
    functionalContext,
    mount,
    shallow
  }))
}

export function functionalContext(context = {}) {
  return {
    context: Object.assign({
      data: {},
      props: {}
    }, context)
  }
}


//requestAnimationFrame polyfill | Milos Djakonovic ( @Miloshio ) | MIT | https://github.com/milosdjakonovic/requestAnimationFrame-polyfill
export function rafPolyfill(w) {
  /**
   *
   * How many times should polyfill call
   * update callback? By canon, it should
   * be 60 times per second, so that ideal
   * framerate 60fps could be reached.
   *
   * However, even native implementations
   * of requestAnimationFrame often cannot
   * do 60fps, but, unlike any polyfill,
   * they can synchronise achievable fps
   * rate with screen refresh rate.
   *
   * So, leave this value 1000/60 unless
   * you target specific browser on spec
   * ific device that is going to work
   * better with custom value. I think
   * that this is the longest comment I've
   * written on single variable so far.
  **/
  var FRAME_RATE_INTERVAL = 1000/60,

  /**
   * All queued callbacks in given cycle.
  **/
  allCallbacks = [],

  executeAllScheduled = false,

  shouldCheckCancelRaf = false,

  /**
   * Callbacks queued for cancellation.
  **/
  callbacksForCancellation = [],

  /**
   * Should callback be cancelled?
   * @param cb - callback
  **/
  isToBeCancelled = function(cb){
    for(var i=0;i<callbacksForCancellation.length;i++){
      if(callbacksForCancellation[i] === cb ){
        callbacksForCancellation.splice(i,1);
        return true;
      }
    }
  },



  /**
   *
   * Executes all (surprise) callbacks in
   * and removes them from callback queue.
   *
  **/
  executeAll = function(){
    executeAllScheduled = false;
    var _allCallbacks = allCallbacks;
    allCallbacks = [];
    for(var i=0;i<_allCallbacks.length;i++){
      if(shouldCheckCancelRaf===true){
        if (isToBeCancelled(_allCallbacks[i])){
          shouldCheckCancelRaf = false;
          return;
        }
      }
      _allCallbacks[i].apply(w, [ new Date().getTime() ] );
    }
  },

  /**
   *
   * requestAnimationFrame polyfill
   * @param callback - callback to be queued & executed | executed
   * @return callback
   *
  **/
  raf = function(callback){
    allCallbacks.push(callback);
    if(executeAllScheduled===false){
      w.setTimeout(executeAll, FRAME_RATE_INTERVAL);
      executeAllScheduled = true;
    }
    return callback;
  },

  /**
   *
   * Cancels raf.
  **/
  cancelRaf = function(callback){
    callbacksForCancellation.push(callback);
    shouldCheckCancelRaf = true;
  },


  //https://gist.github.com/paulirish/1579671
  vendors = ['ms', 'moz', 'webkit', 'o'];

    for(var x = 0; x < vendors.length && !w.requestAnimationFrame; ++x) {
        w.requestAnimationFrame = w[vendors[x]+'RequestAnimationFrame'];
        w.cancelAnimationFrame = w[vendors[x]+'CancelAnimationFrame']
        || w[vendors[x]+'CancelRequestAnimationFrame'];
    }

  if (!w.requestAnimationFrame) w.requestAnimationFrame = raf;
  if (!w.cancelAnimationFrame)  w.cancelAnimationFrame  = cancelRaf;

}
