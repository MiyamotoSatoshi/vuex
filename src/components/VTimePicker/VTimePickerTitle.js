import '../../stylus/components/_time-picker-title.styl'

// Mixins
import PickerButton from '../../mixins/picker-button'

// Utils
import { pad } from '../VDatePicker/util'

/* @vue/component */
export default {
  name: 'v-time-picker-title',

  mixins: [PickerButton],

  props: {
    ampm: Boolean,
    hour: Number,
    minute: Number,
    period: {
      type: String,
      validator: period => period === 'am' || period === 'pm'
    },
    selectingHour: Boolean
  },

  methods: {
    genTime () {
      let hour = this.hour
      if (this.ampm) {
        hour = hour ? ((hour - 1) % 12 + 1) : 12
      }

      const displayedHour = this.hour == null ? '--' : this.ampm ? hour : pad(hour)
      const displayedMinute = this.minute == null ? '--' : pad(this.minute)

      return this.$createElement('div', {
        'class': 'v-time-picker-title__time'
      }, [
        this.genPickerButton('selectingHour', true, displayedHour),
        this.$createElement('span', ':'),
        this.genPickerButton('selectingHour', false, displayedMinute)
      ])
    },
    genAmPm () {
      return this.$createElement('div', {
        staticClass: 'v-time-picker-title__ampm'
      }, [
        this.genPickerButton('period', 'am', 'am'),
        this.genPickerButton('period', 'pm', 'pm')
      ])
    }
  },

  render (h) {
    return h('div', {
      staticClass: 'v-time-picker-title'
    }, [
      this.genTime(),
      this.ampm ? this.genAmPm() : null
    ])
  }
}
