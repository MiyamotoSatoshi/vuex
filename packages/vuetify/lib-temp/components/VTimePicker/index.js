import VTimePicker from './VTimePicker';
import VTimePickerClock from './VTimePickerClock';
import VTimePickerTitle from './VTimePickerTitle';
export { VTimePicker, VTimePickerClock, VTimePickerTitle };
export default {
    $_vuetify_subcomponents: {
        VTimePicker,
        VTimePickerClock,
        VTimePickerTitle
    }
};
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi8uLi9zcmMvY29tcG9uZW50cy9WVGltZVBpY2tlci9pbmRleC5qcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSxPQUFPLFdBQVcsTUFBTSxlQUFlLENBQUE7QUFDdkMsT0FBTyxnQkFBZ0IsTUFBTSxvQkFBb0IsQ0FBQTtBQUNqRCxPQUFPLGdCQUFnQixNQUFNLG9CQUFvQixDQUFBO0FBRWpELE9BQU8sRUFBRSxXQUFXLEVBQUUsZ0JBQWdCLEVBQUUsZ0JBQWdCLEVBQUUsQ0FBQTtBQUUxRCxlQUFlO0lBQ2IsdUJBQXVCLEVBQUU7UUFDdkIsV0FBVztRQUNYLGdCQUFnQjtRQUNoQixnQkFBZ0I7S0FDakI7Q0FDRixDQUFBIiwic291cmNlc0NvbnRlbnQiOlsiaW1wb3J0IFZUaW1lUGlja2VyIGZyb20gJy4vVlRpbWVQaWNrZXInXG5pbXBvcnQgVlRpbWVQaWNrZXJDbG9jayBmcm9tICcuL1ZUaW1lUGlja2VyQ2xvY2snXG5pbXBvcnQgVlRpbWVQaWNrZXJUaXRsZSBmcm9tICcuL1ZUaW1lUGlja2VyVGl0bGUnXG5cbmV4cG9ydCB7IFZUaW1lUGlja2VyLCBWVGltZVBpY2tlckNsb2NrLCBWVGltZVBpY2tlclRpdGxlIH1cblxuZXhwb3J0IGRlZmF1bHQge1xuICAkX3Z1ZXRpZnlfc3ViY29tcG9uZW50czoge1xuICAgIFZUaW1lUGlja2VyLFxuICAgIFZUaW1lUGlja2VyQ2xvY2ssXG4gICAgVlRpbWVQaWNrZXJUaXRsZVxuICB9XG59XG4iXX0=