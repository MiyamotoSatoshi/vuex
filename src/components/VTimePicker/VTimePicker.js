// Components
import VTimePickerTitle from './VTimePickerTitle'
import VTimePickerClock from './VTimePickerClock'

// Mixins
import Picker from '../../mixins/picker'

// Utils
import { createRange } from '../../util/helpers'
import pad from '../VDatePicker/util/pad'

const rangeHours24 = createRange(24)
const rangeHours12am = createRange(12)
const rangeHours12pm = rangeHours12am.map(v => v + 12)
const rangeMinutes = createRange(60)

/* @vue/component */
export default {
  name: 'v-time-picker',

  mixins: [Picker],

  props: {
    allowedHours: Function,
    allowedMinutes: Function,
    format: {
      type: String,
      default: 'ampm',
      validator (val) {
        return ['ampm', '24hr'].includes(val)
      }
    },
    min: String,
    max: String,
    readonly: Boolean,
    scrollable: Boolean,
    value: null
  },

  data () {
    return {
      inputHour: null,
      inputMinute: null,
      period: 'am',
      selectingHour: true
    }
  },

  computed: {
    isAllowedHourCb () {
      if (!this.min && !this.max) return this.allowedHours

      const minHour = this.min ? this.min.split(':')[0] : 0
      const maxHour = this.max ? this.max.split(':')[0] : 23

      return val => {
        return val >= minHour * 1 &&
          val <= maxHour * 1 &&
          (!this.allowedHours || this.allowedHours(val))
      }
    },
    isAllowedMinuteCb () {
      const isHourAllowed = !this.allowedHours || this.allowedHours(this.inputHour)
      if (!this.min && !this.max) {
        return isHourAllowed ? this.allowedMinutes : () => false
      }

      const [minHour, minMinute] = this.min ? this.min.split(':') : [0, 0]
      const [maxHour, maxMinute] = this.max ? this.max.split(':') : [23, 59]
      const minTime = minHour * 60 + minMinute * 1
      const maxTime = maxHour * 60 + maxMinute * 1

      return val => {
        const time = 60 * this.inputHour + val
        return time >= minTime &&
          time <= maxTime &&
          isHourAllowed &&
          (!this.allowedMinutes || this.allowedMinutes(val))
      }
    },
    isAmPm () {
      return this.format === 'ampm'
    }
  },

  watch: {
    value: 'setInputData'
  },

  mounted () {
    this.setInputData(this.value)
  },

  methods: {
    emitValue () {
      if (this.inputHour != null && this.inputMinute != null) {
        this.$emit('input', `${pad(this.inputHour)}:${pad(this.inputMinute)}`)
      }
    },
    setPeriod (period) {
      this.period = period
      if (this.inputHour != null) {
        const newHour = this.inputHour + (period === 'am' ? -12 : 12)
        this.inputHour = this.firstAllowed('hour', newHour)
        this.emitValue()
      }
    },
    setInputData (value) {
      if (value == null) {
        this.inputHour = null
        this.inputMinute = null
        return
      }

      if (value instanceof Date) {
        this.inputHour = value.getHours()
        this.inputMinute = value.getMinutes()
      } else {
        const [, hour, minute, , period] = value.trim().toLowerCase().match(/^(\d+):(\d+)(:\d+)?([ap]m)?$/, '') || []

        this.inputHour = period ? this.convert12to24(parseInt(hour, 10), period) : parseInt(hour, 10)
        this.inputMinute = parseInt(minute, 10)
      }

      this.period = this.inputHour < 12 ? 'am' : 'pm'
    },
    convert24to12 (hour) {
      return hour ? ((hour - 1) % 12 + 1) : 12
    },
    convert12to24 (hour, period) {
      return hour % 12 + (period === 'pm' ? 12 : 0)
    },
    onInput (value) {
      if (this.selectingHour) {
        this.inputHour = this.isAmPm ? this.convert12to24(value, this.period) : value
      } else {
        this.inputMinute = value
      }
      this.emitValue()
    },
    onChange () {
      if (!this.selectingHour) {
        this.$emit('change', this.value)
      } else {
        this.selectingHour = false
      }
    },
    firstAllowed (type, value) {
      const allowedFn = type === 'hour' ? this.isAllowedHourCb : this.isAllowedMinuteCb
      if (!allowedFn) return value

      // TODO: clean up
      const range = type === 'minute'
        ? rangeMinutes
        : (this.isAmPm
          ? (value < 12
            ? rangeHours12am
            : rangeHours12pm)
          : rangeHours24)
      const first = range.find(v => allowedFn((v + value) % range.length + range[0]))
      return ((first || 0) + value) % range.length + range[0]
    },
    genClock () {
      return this.$createElement(VTimePickerClock, {
        props: {
          allowedValues: this.selectingHour ? this.isAllowedHourCb : this.isAllowedMinuteCb,
          color: this.color,
          dark: this.dark,
          double: this.selectingHour && !this.isAmPm,
          format: this.selectingHour ? (this.isAmPm ? this.convert24to12 : val => val) : val => pad(val, 2),
          light: this.light,
          max: this.selectingHour ? (this.isAmPm && this.period === 'am' ? 11 : 23) : 59,
          min: this.selectingHour && this.isAmPm && this.period === 'pm' ? 12 : 0,
          readonly: this.readonly,
          scrollable: this.scrollable,
          size: this.width - ((!this.fullWidth && this.landscape) ? 80 : 20),
          step: this.selectingHour ? 1 : 5,
          value: this.selectingHour ? this.inputHour : this.inputMinute
        },
        on: {
          input: this.onInput,
          change: this.onChange
        },
        ref: 'clock'
      })
    },
    genPickerBody () {
      return this.$createElement('div', {
        staticClass: 'v-time-picker-clock__container',
        style: {
          width: `${this.width}px`,
          height: `${this.width - ((!this.fullWidth && this.landscape) ? 60 : 0)}px`
        },
        key: this.selectingHour
      }, [this.genClock()])
    },
    genPickerTitle () {
      return this.$createElement(VTimePickerTitle, {
        props: {
          ampm: this.isAmPm,
          hour: this.inputHour,
          minute: this.inputMinute,
          period: this.period,
          readonly: this.readonly,
          selectingHour: this.selectingHour
        },
        on: {
          'update:selectingHour': value => (this.selectingHour = value),
          'update:period': this.setPeriod
        },
        ref: 'title',
        slot: 'title'
      })
    }
  },

  render () {
    return this.genPicker('v-picker--time')
  }
}
