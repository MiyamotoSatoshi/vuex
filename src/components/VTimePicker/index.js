import VTimePicker from './VTimePicker'

/* istanbul ignore next */
VTimePicker.install = function install (Vue) {
  Vue.component(VTimePicker.name, VTimePicker)
}

export default VTimePicker
