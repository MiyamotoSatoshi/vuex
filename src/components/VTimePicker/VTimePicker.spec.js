import VTimePicker from '~components/VTimePicker'
import { test } from '~util/testing'

test('VTimePicker.js', ({ mount }) => {
  it('should accept a value', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: '09:00:00'
      }
    })

    expect(wrapper.vm.inputTime).toBe('09:00:00')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should accept a date object for a value', () => {
    const now = new Date('2017-01-01 12:00 AM')
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: now
      }
    })

    expect(wrapper.vm.inputTime).toEqual('12:00am')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should change am/pm when updated from model', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: '9:00am'
      }
    })

    wrapper.setProps({ value: '9:00pm' })

    expect(wrapper.data().period).toBe('pm')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should not change period with 24hr prop', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        format: '24hr',
        value: null
      }
    })

    const ampm = wrapper.instance().inputTime.match(/(am|pm)/)

    expect(ampm).toBe(null)
  })

  it('should set picker to pm when given Date after noon', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: new Date('2017-01-01 12:00 PM')
      }
    })

    expect(wrapper.data().period).toEqual('pm')
  })

  it('should set picker to pm when given string with PM in it', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: '1:00 PM'
      }
    })

    expect(wrapper.data().period).toEqual('pm')
  })

  it('should set picker to pm when given string with pm in it', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: '1:00 pm'
      }
    })

    expect(wrapper.data().period).toEqual('pm')
  })

  it('should set picker to am when given Date before noon', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: new Date('2017-01-01 1:00 AM')
      }
    })

    expect(wrapper.data().period).toEqual('am')
  })

  it('should return proper value for isAllowed method (max)', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        allowedMinutes: [50, 55, 0, 5, 10],
        value: null
      }
    })

    expect([0, 5, 10, 15, 20, 45, 50, 55].map(minute => wrapper.vm.isAllowed('minute', minute)))
      .toEqual([true, true, true, false, false, false, true, true])
  })

  it('should return proper value for isAllowed method (min/max)', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        allowedMinutes: { min: 33, max: 55 },
        value: null
      }
    })

    expect([0, 20, 30, 35, 45, 50, 55].map(minute => wrapper.vm.isAllowed('minute', minute)))
      .toEqual([false, false, false, true, true, true, true])
  })

  it('should return proper value for isAllowed method (function)', () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        allowedMinutes: minute => [55, 0, 5].includes(minute),
        value: null
      }
    })

    expect([0, 5, 45, 50, 55].map(minute => wrapper.vm.isAllowed('minute', minute)))
      .toEqual([true, true, false, false, true])
  })

  it('should reset selectingHour when saved/canceled', async () => {
    const wrapper = mount(VTimePicker, {
      propsData: {
        value: null
      }
    })

    wrapper.vm.selectingHour = false
    wrapper.vm.save()
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.selectingHour).toBe(true)
    wrapper.vm.selectingHour = false
    wrapper.vm.cancel()
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.selectingHour).toBe(true)
  })
})
