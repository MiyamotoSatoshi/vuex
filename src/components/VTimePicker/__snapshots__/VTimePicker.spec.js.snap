// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTimePicker.js should accept a date object for a value 1`] = `

<div class="card picker picker--time picker--time--hours"
     style="height: auto;"
>
  <div class="picker__title">
    <div class="picker--time__title">
      <span class="active">
        12
      </span>
      <span class>
        :00
      </span>
    </div>
    <div>
      <span class="active">
        AM
      </span>
      <span class>
        PM
      </span>
    </div>
  </div>
  <div class="picker__body">
    <div class="picker--time__clock">
      <div class="picker--time__clock-hand hour">
      </div>
      <span class>
        <span>
          1
        </span>
      </span>
      <span class>
        <span>
          2
        </span>
      </span>
      <span class>
        <span>
          3
        </span>
      </span>
      <span class>
        <span>
          4
        </span>
      </span>
      <span class>
        <span>
          5
        </span>
      </span>
      <span class>
        <span>
          6
        </span>
      </span>
      <span class>
        <span>
          7
        </span>
      </span>
      <span class>
        <span>
          8
        </span>
      </span>
      <span class>
        <span>
          9
        </span>
      </span>
      <span class>
        <span>
          10
        </span>
      </span>
      <span class>
        <span>
          11
        </span>
      </span>
      <span class="active">
        <span>
          12
        </span>
      </span>
    </div>
  </div>
</div>

`;

exports[`VTimePicker.js should accept a value 1`] = `

<div class="card picker picker--time picker--time--hours"
     style="height: auto;"
>
  <div class="picker__title">
    <div class="picker--time__title">
      <span class="active">
        9
      </span>
      <span class>
        :00
      </span>
    </div>
    <div>
      <span class="active">
        AM
      </span>
      <span class>
        PM
      </span>
    </div>
  </div>
  <div class="picker__body">
    <div class="picker--time__clock">
      <div class="picker--time__clock-hand hour">
      </div>
      <span class>
        <span>
          1
        </span>
      </span>
      <span class>
        <span>
          2
        </span>
      </span>
      <span class>
        <span>
          3
        </span>
      </span>
      <span class>
        <span>
          4
        </span>
      </span>
      <span class>
        <span>
          5
        </span>
      </span>
      <span class>
        <span>
          6
        </span>
      </span>
      <span class>
        <span>
          7
        </span>
      </span>
      <span class>
        <span>
          8
        </span>
      </span>
      <span class="active">
        <span>
          9
        </span>
      </span>
      <span class>
        <span>
          10
        </span>
      </span>
      <span class>
        <span>
          11
        </span>
      </span>
      <span class>
        <span>
          12
        </span>
      </span>
    </div>
  </div>
</div>

`;

exports[`VTimePicker.js should change am/pm when updated from model 1`] = `

<div class="card picker picker--time picker--time--hours"
     style="height: auto;"
>
  <div class="picker__title">
    <div class="picker--time__title">
      <span class="active">
        9
      </span>
      <span class>
        :00
      </span>
    </div>
    <div>
      <span class="active">
        AM
      </span>
      <span class>
        PM
      </span>
    </div>
  </div>
  <div class="picker__body">
    <div class="picker--time__clock">
      <div class="picker--time__clock-hand hour">
      </div>
      <span class>
        <span>
          1
        </span>
      </span>
      <span class>
        <span>
          2
        </span>
      </span>
      <span class>
        <span>
          3
        </span>
      </span>
      <span class>
        <span>
          4
        </span>
      </span>
      <span class>
        <span>
          5
        </span>
      </span>
      <span class>
        <span>
          6
        </span>
      </span>
      <span class>
        <span>
          7
        </span>
      </span>
      <span class>
        <span>
          8
        </span>
      </span>
      <span class="active">
        <span>
          9
        </span>
      </span>
      <span class>
        <span>
          10
        </span>
      </span>
      <span class>
        <span>
          11
        </span>
      </span>
      <span class>
        <span>
          12
        </span>
      </span>
    </div>
  </div>
</div>

`;
